From ce44dd0ba41b88f6b8eb626863c3358b75aca66c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 2 Dec 2025 12:32:28 +0100
Subject: [PATCH 1/3] wined3d: Remove now unnecessary pixel format restoration.

wglSetPixelFormatWINE now sets the internal pixel format privately on
each DC, and it is unnecessary to restore the window pixel format.

That call is even undesirable as it allocates and destroys a private
OpenGL drawable when the DC is acquired, its internal pixel format is
set, and the DC is then immediately released.
---
 dlls/wined3d/context_gl.c | 44 ++-------------------------------------
 dlls/wined3d/wined3d_gl.h |  1 -
 2 files changed, 2 insertions(+), 43 deletions(-)

diff --git a/dlls/wined3d/context_gl.c b/dlls/wined3d/context_gl.c
index ac1b9370fc1..82e44161598 100644
--- a/dlls/wined3d/context_gl.c
+++ b/dlls/wined3d/context_gl.c
@@ -1162,38 +1162,6 @@ void wined3d_context_gl_texture_update(struct wined3d_context_gl *context_gl,
     }
 }
 
-static BOOL wined3d_context_gl_restore_pixel_format(struct wined3d_context_gl *context_gl)
-{
-    const struct wined3d_gl_info *gl_info = context_gl->gl_info;
-    BOOL ret = FALSE;
-
-    if (context_gl->restore_pf && IsWindow(context_gl->restore_pf_win))
-    {
-        if (gl_info->supported[WGL_WINE_PIXEL_FORMAT_PASSTHROUGH])
-        {
-            HDC dc = GetDCEx(context_gl->restore_pf_win, 0, DCX_USESTYLE | DCX_CACHE);
-            if (dc)
-            {
-                if (!(ret = GL_EXTCALL(wglSetPixelFormatWINE(dc, context_gl->restore_pf))))
-                {
-                    ERR("Failed to restore pixel format %d on window %p.\n",
-                            context_gl->restore_pf, context_gl->restore_pf_win);
-                }
-                ReleaseDC(context_gl->restore_pf_win, dc);
-            }
-        }
-        else
-        {
-            ERR("Unable to restore pixel format %d on window %p.\n",
-                    context_gl->restore_pf, context_gl->restore_pf_win);
-        }
-    }
-
-    context_gl->restore_pf = 0;
-    context_gl->restore_pf_win = NULL;
-    return ret;
-}
-
 static BOOL wined3d_context_gl_set_pixel_format(struct wined3d_context_gl *context_gl)
 {
     const struct wined3d_gl_info *gl_info = context_gl->gl_info;
@@ -1201,7 +1169,6 @@ static BOOL wined3d_context_gl_set_pixel_format(struct wined3d_context_gl *conte
     int format = context_gl->pixel_format;
     HDC dc = context_gl->dc;
     int current;
-    HWND win;
 
     if (private && context_gl->dc_has_format)
         return TRUE;
@@ -1246,12 +1213,6 @@ static BOOL wined3d_context_gl_set_pixel_format(struct wined3d_context_gl *conte
         return FALSE;
     }
 
-    win = private ? NULL : WindowFromDC(dc);
-    if (win != context_gl->restore_pf_win)
-        wined3d_context_gl_restore_pixel_format(context_gl);
-    context_gl->restore_pf = private ? 0 : current;
-    context_gl->restore_pf_win = win;
-
 success:
     if (private)
         context_gl->dc_has_format = TRUE;
@@ -1536,7 +1497,6 @@ static void wined3d_context_gl_cleanup(struct wined3d_context_gl *context_gl)
 
     free(context_gl->texture_type);
 
-    wined3d_context_gl_restore_pixel_format(context_gl);
     if (restore_ctx)
         context_restore_gl_context(gl_info, restore_dc, restore_ctx);
     else if (wglGetCurrentContext() && !wglMakeCurrent(NULL, NULL))
@@ -1642,8 +1602,8 @@ void wined3d_context_gl_release(struct wined3d_context_gl *context_gl)
 
     if (!--context_gl->level)
     {
-        if (wined3d_context_gl_restore_pixel_format(context_gl))
-            context_gl->needs_set = 1;
+        context_gl->needs_set = 1;
+
         if (context_gl->restore_ctx)
         {
             TRACE("Restoring GL context %p on device context %p.\n", context_gl->restore_ctx, context_gl->restore_dc);
diff --git a/dlls/wined3d/wined3d_gl.h b/dlls/wined3d/wined3d_gl.h
index 0cb36eadafe..25174caaf8e 100644
--- a/dlls/wined3d/wined3d_gl.h
+++ b/dlls/wined3d/wined3d_gl.h
@@ -647,7 +647,6 @@ struct wined3d_context_gl
     HGLRC restore_ctx;
     HDC restore_dc;
     int restore_pf;
-    HWND restore_pf_win;
     HGLRC gl_ctx;
     HDC dc;
     int pixel_format;
-- 
2.52.0


From f80159a1448f56b315230dc370ff1fc123b8b7fd Mon Sep 17 00:00:00 2001
From: Henri Verbeet <hverbeet@locutus.nl>
Date: Sat, 6 Dec 2025 20:01:59 +0100
Subject: [PATCH 2/3] wined3d: Do not set "context_gl->needs_set" in
 wined3d_context_gl_release().

Commit ce44dd0ba41b88f6b8eb626863c3358b75aca66c got rid of
wined3d_context_gl_restore_pixel_format(), but seemingly inverted the
condition under which wined3d_context_gl_release() set the "needs_set"
flag. Previously the flag was set when
wined3d_context_gl_restore_pixel_format() changed the pixel format, and
now it's set when it doesn't. I.e., always.

Setting the "needs_set" flag causes wined3d_context_gl_activate() to
call wined3d_context_gl_set_gl_context(), which sets the pixel format and
makes the GL context current. These are expensive operations. This
commit improves performance in the rthdribl demo from a fairly
disappointing 1-2 fps to a more reasonable 70-80 fps on my Intel Skylake
setup. On my AMD gfx1200 setup the difference is less dramatic but still
quite significant, going from about 250 fps to about 350 fps.

It's perhaps worth pointing out that after this commit things are still
some way off from what they were on Wine 10.0: 150-160 fps on the Intel
setup, and about 450 fps on the AMD setup.
---
 dlls/wined3d/context_gl.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/wined3d/context_gl.c b/dlls/wined3d/context_gl.c
index 729f4e784bb..6c1b169f4f4 100644
--- a/dlls/wined3d/context_gl.c
+++ b/dlls/wined3d/context_gl.c
@@ -1591,8 +1591,6 @@ void wined3d_context_gl_release(struct wined3d_context_gl *context_gl)
 
     if (!--context_gl->level)
     {
-        context_gl->needs_set = 1;
-
         if (context_gl->restore_ctx)
         {
             TRACE("Restoring GL context %p on device context %p.\n", context_gl->restore_ctx, context_gl->restore_dc);
-- 
2.52.0


From e388e55ce5238bc03200be92d61b81f972e97741 Mon Sep 17 00:00:00 2001
From: Henri Verbeet <hverbeet@locutus.nl>
Date: Sat, 6 Dec 2025 20:15:23 +0100
Subject: [PATCH 3/3] wined3d: Get rid of the "restore_pf" field from struct
 wined3d_context_gl.

Its usage was removed in commit ce44dd0ba41b88f6b8eb626863c3358b75aca66c.
---
 dlls/wined3d/wined3d_gl.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/wined3d/wined3d_gl.h b/dlls/wined3d/wined3d_gl.h
index 25174caaf8e..2c22c0a9c3c 100644
--- a/dlls/wined3d/wined3d_gl.h
+++ b/dlls/wined3d/wined3d_gl.h
@@ -646,7 +646,6 @@ struct wined3d_context_gl
     unsigned int level;
     HGLRC restore_ctx;
     HDC restore_dc;
-    int restore_pf;
     HGLRC gl_ctx;
     HDC dc;
     int pixel_format;
-- 
2.52.0

